File: app/src/main/java/com/mandarin/bcu/androidutil/lineup/adapters/ComboSubSchListAdapter.java
package com.mandarin.bcu.androidutil.lineup.adapters;

import android.app.Activity;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.mandarin.bcu.R;
import com.mandarin.bcu.androidutil.StaticStore;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import common.system.MultiLangCont;
import common.util.unit.Combo;

public class ComboSubSchListAdapter extends ArrayAdapter<String> {
    private List<String> sch;
    private boolean [] combocat;
    private ListView combolist;
    private Activity activity;
    private ComboListAdapter comboListAdapter;

    private List<String> comid = new ArrayList<>();
    private List<Integer> defcom;

    ComboSubSchListAdapter(Activity activity, List<String> sch, ListView combolist, List<Integer> defcom, ComboListAdapter comboListAdapter) {
        super(activity, R.layout.spinneradapter,sch);
        this.sch = sch;
        this.combolist = combolist;
        this.activity = activity;
        this.defcom = defcom;
        this.comboListAdapter = comboListAdapter;

        combocat = new boolean[sch.size()];
    }

    private static class ViewHolder {
        TextView category;

        private ViewHolder(View view) {
            category = view.findViewById(R.id.spinnertext);
        }
    }

    @NonNull
    public View getView(int position, View view, @NonNull ViewGroup parent) {
        View row;
        ViewHolder holder;

        if(view == null) {
            LayoutInflater inflater = LayoutInflater.from(getContext());
            row = inflater.inflate(R.layout.spinneradapter,parent,false);
            holder = new ViewHolder(row);
            row.setTag(holder);
        } else {
            row = view;
            holder = (ViewHolder) row.getTag();
        }

        holder.category.setText(sch.get(position));

        holder.category.setBackgroundColor(combocat[position] ?StaticStore.getAttributeColor(getContext(),R.attr.SelectionPrimary):getContext().getColor(android.R.color.transparent));

        holder.category.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                combocat[position] = !combocat[position];
                holder.category.setBackgroundColor(combocat[position] ?StaticStore.getAttributeColor(getContext(),R.attr.SelectionPrimary):getContext().getColor(android.R.color.transparent));

                if(combocat[position]) {
                    comid.add(String.valueOf(defcom.get(position)));
                } else {
                    comid.remove(String.valueOf(defcom.get(position)));
                }

                StaticStore.combos.clear();

                if(comid.isEmpty()) {
                    for(int i = 0; i < defcom.size(); i++) {
                        StaticStore.combos.addAll(Arrays.asList(Combo.combos[defcom.get(i)]));
                    }

                    String [] names = new String[StaticStore.combos.size()];

                    for(int i = 0; i < StaticStore.combos.size(); i++) {
                        names[i] = MultiLangCont.COMNAME.getCont(StaticStore.combos.get(i).name);
                    }

                    comboListAdapter = new ComboListAdapter(activity,names);

                    combolist.setAdapter(comboListAdapter);
                } else {

                    for(int i = 0; i < comid.size(); i++) {
                        StaticStore.combos.addAll(Arrays.asList(Combo.combos[Integer.parseInt(comid.get(i))]));
                    }

                    String [] names = new String[StaticStore.combos.size()];

                    for(int i = 0; i < StaticStore.combos.size(); i++) {
                        names[i] = MultiLangCont.COMNAME.getCont(StaticStore.combos.get(i).name);
                    }

                    comboListAdapter = new ComboListAdapter(activity,names);

                    combolist.setAdapter(comboListAdapter);
                }
            }
        });

        return row;
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/lineup/adapters/ComboSubSchListAdapter.kt
package com.mandarin.bcu.androidutil.lineup.adapters

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.TextView
import com.mandarin.bcu.R
import com.mandarin.bcu.androidutil.StaticStore
import common.pack.Identifier
import common.pack.UserProfile
import common.util.lang.MultiLangCont
import common.util.unit.Combo

class ComboSubSchListAdapter internal constructor(private val activity: Activity, private val sch: List<String?>, private val combolist: ListView, private val defcom: List<Int>, private var comboListAdapter: ComboListAdapter?) : ArrayAdapter<String?>(activity, R.layout.spinneradapter, sch) {
    private val combocat: BooleanArray = BooleanArray(sch.size)
    private val comid = ArrayList<Int>()

    private class ViewHolder constructor(view: View) {
        var category: TextView = view.findViewById(R.id.spinnertext)

    }

    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
        val holder: ViewHolder
        val row: View

        if(view == null) {
            val inf = LayoutInflater.from(context)

            row = inf.inflate(R.layout.spinneradapter,parent,false)

            holder = ViewHolder(row)

            row.tag = holder

        } else {
            row = view

            holder = row.tag as ViewHolder
        }

        holder.category.text = sch[position]

        holder.category.setBackgroundColor(
            if (combocat[position])
                StaticStore.getAttributeColor(context, R.attr.SelectionPrimary)
            else
                context.getColor(android.R.color.transparent)
        )

        holder.category.setOnClickListener {
            combocat[position] = !combocat[position]

            holder.category.setBackgroundColor(
                if (combocat[position])
                    StaticStore.getAttributeColor(context, R.attr.SelectionPrimary)
                else
                    context.getColor(android.R.color.transparent)
            )

            if (combocat[position]) {
                comid.add(defcom[position])
            } else {
                comid.remove(defcom[position])
            }

            StaticStore.combos.clear()

            if (comid.isEmpty()) {
                StaticStore.combos.addAll(UserProfile.getBCData().combos.list)

                for(userPack in UserProfile.getUserPacks()) {
                    for(combo in userPack.combos.list) {
                        combo ?: continue

                        StaticStore.combos.add(combo)
                    }
                }

                StaticStore.combos.sortWith(Comparator.comparingInt(Combo::type).thenComparingInt(Combo::lv))

                val names = Array<String>(StaticStore.combos.size) {
                    if(StaticStore.combos[it].id.pack == Identifier.DEF) {
                        MultiLangCont.getStatic().COMNAME.getCont(StaticStore.combos[it])
                    } else {
                        StaticStore.combos[it].name
                    }
                }

                comboListAdapter = ComboListAdapter(activity, names)
                combolist.adapter = comboListAdapter
            } else {
                for(pack in UserProfile.getAllPacks()) {
                    for(c in pack.combos.list) {
                        c ?: continue

                        for(l in comid) {
                            if(c.type == l)
                                StaticStore.combos.add(c)
                        }
                    }
                }

                StaticStore.combos.sortWith(Comparator.comparingInt(Combo::type).thenComparingInt(Combo::lv))

                val names = Array<String>(StaticStore.combos.size) {
                    if(StaticStore.combos[it].id.pack == Identifier.DEF) {
                        MultiLangCont.getStatic().COMNAME.getCont(StaticStore.combos[it])
                    } else {
                        StaticStore.combos[it].name
                    }
                }

                comboListAdapter = ComboListAdapter(activity, names)
                combolist.adapter = comboListAdapter
            }
        }
        return row
    }

}